src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,31 +12,92 @@ function calc_error_norms(func, u, t, analyzer,
   # TODO: MPI, we should improve this; maybe we should dispatch on `u`
   #       and create some MPI array type, overloading broadcasting and mapreduce etc.
   #       Then, this specific array type should also work well with DiffEq etc.
-  l2_error, linf_error = invoke(calc_error_norms,
-    Tuple{typeof(func), typeof(u), typeof(t), typeof(analyzer), TreeMesh{2},
-          typeof(equations), typeof(initial_condition), typeof(dg), typeof(cache),
-          typeof(cache_analysis)},
-    func, u, t, analyzer, mesh, equations, initial_condition, dg, cache, cache_analysis)
-
-  # Since the local L2 norm is already normalized and square-rooted, we need to undo this first
-  total_volume = mesh.tree.length_level_0^ndims(mesh)
-  global_l2_error = Vector(l2_error.^2 .* total_volume)
-  global_linf_error = Vector(linf_error)
-  MPI.Reduce!(global_l2_error, +, mpi_root(), mpi_comm())
-  MPI.Reduce!(global_linf_error, max, mpi_root(), mpi_comm())
+  l2_errors, linf_errors = calc_error_norms_per_element(func, u, t, analyzer,
+                                                        mesh, equations, initial_condition,
+                                                        dg, cache, cache_analysis)
+
+  nvars = length(l2_errors[1])
+  T = eltype(l2_errors)
+
+  # Convert errors from Vector of SVectors to Matrix for MPI communication
+  FT = eltype(l2_errors[1])
+  l2_errors = reshape(reinterpret(FT, l2_errors), nvars, :) |> Matrix
+  linf_errors = reshape(reinterpret(FT, linf_errors), nvars, :) |> Matrix
+
+  # Collect local error norms of all elements on root process
   if mpi_isroot()
-    l2_error   = convert(typeof(l2_error),   global_l2_error)
-    linf_error = convert(typeof(linf_error), global_linf_error)
+    global_l2_errors = zeros(eltype(l2_errors), nvars, cache.mpi_cache.n_elements_global)
+    global_linf_errors = similar(global_l2_errors)
```

Comment:
That doesn't scale, does it? Collecting information from all elements on the root process?

Reply:
No, technically it does not scale. However, error calculation is inherently a costly process, and even for 3D compressible Euler and 1 million elements (which I don't think we will reach anytime soon), this amounts to just ~40 MB. Thus I'd argue that thinking about scalability here is for now like premature optimization. We could, however, add a comment to warn about this?

Reply:
Added a comment in 30abccf that mentions that it doesn't scale.

Reply:
Thanks!